packages/core/src/MachineNode.ts

```diff
@@ -0,0 +1,351 @@
+
+  public __xstatenode: true = true;
+
+  public __cache = {
+    events: undefined as Array<TEvent['type']> | undefined,
+    relativeValue: new Map() as Map<StateNode<TContext>, StateValue>,
+    initialStateValue: undefined as StateValue | undefined,
+    initialState: undefined as State<TContext, TEvent> | undefined,
+    on: undefined as TransitionDefinitionMap<TContext, TEvent> | undefined,
+    transitions: undefined as
+      | Array<TransitionDefinition<TContext, TEvent>>
+      | undefined,
+    candidates: {} as {
+      [K in TEvent['type'] | NullEvent['type'] | '*']:
+        | Array<
+            TransitionDefinition<
+              TContext,
+              K extends TEvent['type']
+                ? Extract<TEvent, { type: K }>
+                : EventObject
+            >
+          >
+        | undefined;
+    },
+    delayedTransitions: undefined as
+      | Array<DelayedTransitionDefinition<TContext, TEvent>>
+      | undefined
+  };
+
+  constructor(
+    /**
+     * The raw config used to create the machine.
+     */
+    public config: MachineConfig<TContext, TStateSchema, TEvent>,
+    options?: Partial<MachineOptions<TContext, TEvent>>
+  ) {
+    super(config, {
+      _key: config.id || '(machine)'
+    });
+    this.options = Object.assign(
+      createDefaultOptions(config.context!),
+      options
+    );
+    this.context = resolveContext<TContext>(
+      config.context as TContext,
+      this.options.context
+    );
+    this.key = this.config.key || this.config.id || '(machine)';
+    this.machine = this;
+    this.path = [];
+    this.delimiter = this.config.delimiter || STATE_DELIMITER;
+    this.version = this.config.version;
+
+    // Document order
+    let order = 0;
+
+    function dfs(stateNode: StateNode<TContext, any, TEvent>): void {
+      stateNode.order = order++;
+
+      for (const child of getChildren(stateNode)) {
+        dfs(child);
+      }
+    }
+
+    dfs(this);
+
+    this.strict = !!this.config.strict;
+
+    this.entry = toArray(this.config.entry).map(action =>
+      toActionObject(action)
+    );
+
+    this.exit = toArray(this.config.exit).map(action => toActionObject(action));
+    this.meta = this.config.meta;
+    this.transition = this.transition.bind(this);
+  }
+
+  private _init(): void {
+    if (this.__cache.transitions) {
+      return;
+    }
+    getAllStateNodes(this).forEach(stateNode => stateNode.on);
+  }
+
+  /**
+   * Clones this state machine with custom options and context.
+   *
+   * @param options Options (actions, guards, activities, services) to recursively merge with the existing options.
+   * @param context Custom context (will override predefined context)
+   */
+  public withConfig(
+    options: Partial<MachineOptions<TContext, TEvent>>
+  ): MachineNode<TContext, TStateSchema, TEvent> {
+    const {
+      actions,
+      activities,
+      guards,
+      services,
+      delays,
+      context = this.context
```

Comment:
I believe this has a slight bug - we can have a machine config with a defined context and create a machine based on it with context option. If we call `withConfig` on it later we might end up with an incorrect value of `context` here because it will be the one from `this.options.context`, but it should actually always just point to `this.context` (which is a resolved context).

I would just recommend removing this line from here and use `this.context` directly when calling MachineNode constructor those few lines below.

Reply:
If my hunch was right about this being a slight bug then it would be good to add a test covering this.

Reply:
The line was removed (this code is outdated) - does this still apply?

Reply:
I think the code before the change had this slight bug, whereas the new version doesnt have it - so it has fixed the issue, but can potentially regress in the future if not covered by a test.